
<launch>
   <!-- Backward compatibility launch file, use "rtabmap.launch rgbd:=false stereo:=true" instead -->

   <!-- Your camera should be calibrated and publishing rectified left and right
        images + corresponding camera_info msgs. You can use stereo_image_proc for image rectification.
        Example:
           $ roslaunch rtabmap_ros bumblebee.launch -->

  <!-- Choose visualization -->
  <arg name="rtabmapviz"              default="false" />
  <arg name="rviz"                    default="false" />

  <!-- Localization-only mode -->
  <arg name="localization"            default="false"/>

  <arg name="frame_id"                default="xiroi"/>     <!-- Fixed frame id, you may set "base_link" or "base_footprint" if they are published -->
  <arg name="database_path"           default="~/.ros/rtabmap.db"/>
  <arg name="rtabmap_args"            default=""/>   <!-- delete_db_on_start, udebug -->
  <arg name="odom_args"               default=""/>              <!-- More arguments for odometry (overwrite same parameters in rtabmap_args) -->
  <arg name="launch_prefix"           default=""/>
  <arg name="approx_sync"             default="true"/>         <!-- if timestamps of the input topics are not synchronized -->

  <!-- Corresponding config files -->
  <arg name="cfg"                     default="" /> <!-- To change RTAB-Map's parameters, set the path of config file (*.ini) generated by the standalone app -->

  <arg name="stereo_namespace"        default="/stereo_down/scaled_x2"/>
  <arg name="left_image_topic"        default="$(arg stereo_namespace)/left/image_rect_color" />
  <arg name="right_image_topic"       default="$(arg stereo_namespace)/right/image_rect" />      <!-- using grayscale image for efficiency -->
  <arg name="left_camera_info_topic"  default="$(arg stereo_namespace)/left/camera_info" />
  <arg name="right_camera_info_topic" default="$(arg stereo_namespace)/right/camera_info" />
  <arg name="compressed"              default="false"/>

  <arg name="subscribe_scan"          default="false"/>         <!-- Assuming 2D scan if set, rtabmap will do 3DoF mapping instead of 6DoF -->
  <arg name="scan_topic"              default="/scan"/>

  <arg name="subscribe_scan_cloud"    default="false"/>         <!-- Assuming 3D scan if set -->
  <arg name="scan_cloud_topic"        default="/scan_cloud"/>

  <arg name="visual_odometry"         default="true"/>          <!-- Generate visual odometry -->
  <arg name="odom_topic"              default="/odom"/>         <!-- Odometry topic used if visual_odometry is false -->
  <arg name="odom_frame_id"           default="odom"/>              <!-- If set, TF is used to get odometry instead of the topic -->
  <arg name="map_frame_id"            default="map"/>

  <arg name="namespace"               default="rtabmap"/>
  <arg name="queue_size"              default="10"/>
  <arg name="wait_for_transform"      default="0.2"/>
  <arg name="stereo"                  default="true"/>
  <arg name="output"                  default="screen"/>        <!-- Control node output (screen or log) -->

  <!-- Nodes -->
  <group ns="$(arg namespace)">
    <!-- Stereo Odometry -->
    <node if="$(arg stereo)" pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry" output="$(arg output)" args="$(arg rtabmap_args) $(arg odom_args)" launch-prefix="$(arg launch_prefix)">
      <remap from="left/image_rect"        to="$(arg left_image_topic)"/>
      <remap from="right/image_rect"       to="$(arg right_image_topic)"/>
      <remap from="left/camera_info"       to="$(arg left_camera_info_topic)"/>
      <remap from="right/camera_info"      to="$(arg right_camera_info_topic)"/>
      <remap from="odom"                   to="$(arg odom_topic)"/>

      <param name="frame_id"                    type="string" value="$(arg frame_id)"/>
      <param name="odom_frame_id"               type="string" value="$(arg odom_frame_id)"/>

      <param name="wait_for_transform_duration" type="double" value="$(arg wait_for_transform)"/>
      <param name="approx_sync" type="bool" value="false"/>
      <param name="queue_size" type="int" value="5"/>

      <param name="config_path"                 type="string" value="$(arg cfg)"/>
      <param name="queue_size"                  type="int"    value="$(arg queue_size)"/>
      <param name="subscribe_rgbd"              type="bool"   value="false"/>
      <param name="guess_frame_id"              type="string" value=""/>
      <param name="guess_min_translation"       type="double" value="0"/>
      <param name="guess_min_rotation"          type="double" value="0"/>

      <param name="Odom/Strategy" type="string" value="1"/> <!-- 0=Frame-to-Map (F2M) 1=Frame-to-Frame (F2F) 2=Fovis 3=viso2 4=DVO-SLAM 5=ORB_SLAM2 -->
      <!-- maximum features map size, default 2000 -->
      <param name="OdomF2M/MaxSize" type="string" value="1000"/>
      <!-- maximum features extracted by image, default 1000 -->
      <param name="Vis/MaxFeatures" type="string" value="600"/>

      <param name="Vis/EstimationType" type="string" value="1"/> <!-- 3D->2D (PnP) -->
      <param name="Vis/MinInliers" type="string" value="10"/>
      <param name="Vis/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
      <param name="Vis/MaxDepth" type="string" value="10"/>
      <param name="Vis/CorNNDR" type="string" value="0.8"/>
      <param name="Odom/FillInfoData" type="string" value="$(arg rtabmapviz)"/>
      <param name="GFTT/MinDistance" type="string" value="10"/>
      <param name="GFTT/QualityLevel" type="string" value="0.00001"/>

      <param name="Reg/Force3DoF" type="bool" value="true"/>
    </node>


    <!-- Visual SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="$(arg output)" args="$(arg rtabmap_args)" launch-prefix="$(arg launch_prefix)">
      <param name="frame_id" type="string" value="$(arg frame_id)"/>
      <param name="subscribe_stereo" type="bool" value="true"/>
      <param name="subscribe_depth" type="bool" value="false"/>
      <param name="stereo_approx_sync" type="bool" value="false"/>

      <param name="queue_size" type="int" value="30"/>

      <remap from="left/image_rect"        to="$(arg left_image_topic)"/>
      <remap from="right/image_rect"       to="$(arg right_image_topic)"/>
      <remap from="left/camera_info"       to="$(arg left_camera_info_topic)"/>
      <remap from="right/camera_info"      to="$(arg right_camera_info_topic)"/>

      <remap from="odom"                   to="$(arg odom_topic)"/>

      <!-- localization mode -->
      <param name="Mem/IncrementalMemory" type="string" value="true"/>
      <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>

      <!-- RTAB-Map's parameters -->
      <param name="Rtabmap/TimeThr" type="string" value="700"/>
      <param name="Rtabmap/DetectionRate" type="string" value="1"/>

      <param name="Kp/MaxFeatures" type="string" value="200"/>
      <param name="Kp/NNStrategy" type="string" value="0.8"/>         <!-- NNDR ratio (A matching pair is detected, if its distance is closer than X times the distance of the second nearest neighbor. -->
      <param name="Kp/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/> <!-- Region of interest ratios [left, right, top, bottom]. -->
      <param name="Kp/DetectorStrategy" type="string" value="9"/>   <!-- 0=SURF 1=SIFT 2=ORB 3=FAST/FREAK 4=FAST/BRIEF 5=GFTT/FREAK 6=GFTT/BRIEF 7=BRISK 8=GFTT/ORB 9=KAZE. -->

      <param name="Reg/Force3DoF" type="bool" value="true"/>

      <param name="Vis/MinInliers" type="string" value="15"/>
      <param name="Vis/EstimationType" type="string" value="1"/> <!-- 3D->2D (PnP) -->
      <param name="Vis/BundleAdjustment" type="bool" value="true"/>
    </node>
  </group>

</launch>
